### ISSUE TITLE
transfer configuration / pi-manage full config import/export

### ISSUE THREAD

**Issue by laclaro** [privacyidea/privacyidea#2455]
# Configuration transfer

pi-manage should be able to export and import the complete configuration to duplicate the configuration e.g. to duplicate the productive setup for development and testing purposes.

- [x] Implementation of the base functionality to export and import configurations https://github.com/privacyidea/privacyidea/pull/2696
- Config sections to be exported
   - [x] ca connectors https://github.com/privacyidea/privacyidea/pull/2822/commits/faa5c79a35f19ea7886cb95a06617333199ae124
   - [x] system config https://github.com/privacyidea/privacyidea/pull/2696
   - [x] events https://github.com/privacyidea/privacyidea/pull/2822
   - [x] machine resolvers https://github.com/privacyidea/privacyidea/pull/2822/commits/faa5c79a35f19ea7886cb95a06617333199ae124
   - [x] periodic tasks https://github.com/privacyidea/privacyidea/pull/2822/commits/faa5c79a35f19ea7886cb95a06617333199ae124
   - [x] policies https://github.com/privacyidea/privacyidea/pull/2822
   - [x] privacyidea servers https://github.com/privacyidea/privacyidea/pull/2822/commits/32567e80adb0d571236eb9a5663ed1c64b96fec3
   - [x] radius servers https://github.com/privacyidea/privacyidea/pull/2822/commits/bb5b61e491dfaa50a94bc7a8f033e9c382ce848f
   - [x] realms https://github.com/privacyidea/privacyidea/pull/2696
   - [x] resolvers https://github.com/privacyidea/privacyidea/pull/2696
   - [x] sms gateways https://github.com/privacyidea/privacyidea/pull/2822/commits/6862783b31abf43834ee65de006203f91d7960fe
   - [x] smtp servers https://github.com/privacyidea/privacyidea/pull/2822/commits/da4bc32cf90804e3db65fbf5e1a8a94b58595635 and https://github.com/privacyidea/privacyidea/pull/2822/commits/57d517f8fd5faaaaa1007e3fa75131a0f984613b

The actual data is not exported
- internal admins
- tokens
- token owners/assignments
- subscriptions
- audit
- stats

_Original issue text_
> In some cases we would like to transfer a privacyidea database from one development environment to another. Maybe these
> environments have different MySQL database connections (ip, username password, db-name).
>
> This could be done if pi-manage would accept and use a connection string parameter in its backup restore. To do this, only the present connection string in pi.cfg must be replaced by the given one.

**Comment by cornelinux** [comment]
Come and talk to me.

**Comment by cornelinux** [comment]
1. The pi-manage does not work with different database types anyways.

2. If you move the data to another system, pi-manage will read the pi.cfg of the other system anyways, since you would scp the backup to the other machine *locally!*

So I do not see the use of this?

**Comment by laclaro** [comment]
Ah, that's correct. I only refer to mysql in this case.

But I think in 2. you are mistaken. pi-manage uses the username and password, it finds in the backup pi.cfg:

https://github.com/privacyidea/privacyidea/blob/6a35b717b8aebe5bcd2b1f71d7182a8755c610c6/pi-manage#L385

This fails if you move to a different machine.

Maybe one should only change this behavior, so that the restore uses the connection string of the target machine.

If we do that, we should make the restore a little more intelligent and also compare the hashes of backup pi.cfg and target pi.cfg and warn if they are different.

Maybe we should change the issue subject to "make pi-manage restore more intelligent".

**Comment by cornelinux** [comment]
Yes, the pi.cfg is overwritten from the backup.  But I still do not see the problem.
What do you want to achieve?
Let's talk about this. Please.

**Comment by laclaro** [comment]
We concluded that the task I have in mind is more a configuration transfer task rather than a backup restore task.
If you would like to duplicate your privacyIDEA configuration with a new machine it is unlikely that you also want to duplicate your token/audit data.

We will enhance the export/import functionality of pi-manage to be able to do a configuration transfer (different topic).

# Enhance pi-manage backup restore

- pi-manage backup restore should do thing or make this behavior configurable
   1. use the username/password from the *existing* target installation of privacyIDEA *or*
   2. setup a new database with name as specified and grant permissions to the specified user with the specified password. This will fail if we do not have full access to the mysql database.

**Comment by cornelinux** [comment]
@laclaro Please check what is actually missing to do a full export on machineA and import it on machineB. Also: Check, when importing - do we cleanup old config or **add** it to existing config - or allow both.
First add a checklist to this issue.

**Comment by laclaro** [comment]
- [ ] ca connectors
- [ ] general config
- [ ] events
- [ ] machine resolvers
- [ ] periodic tasks
- [ ] policies
- [ ] privacyidea servers
- [ ] radius servers
- [ ] realms
- [ ] resolvers
- [ ] sms gateways
- [ ] smtp servers

The actual data is not exported
- internal admins
- tokens
- token owners/assignments
- subscriptions
- audit
- stats

**Comment by laclaro** [comment]
# Configuration transfer

pi-manage should be able to export and import the complete configuration to duplicate the configuration e.g. to duplicate the productive setup for development and testing purposes.

Config sections to be exported
- [ ] ca connectors
- [ ] general config
- [x] events
- [ ] machine resolvers
- [ ] periodic tasks
- [x] policies
- [ ] privacyidea servers
- [ ] radius servers
- [ ] realms
- [x] resolvers
- [ ] sms gateways
- [ ] smtp servers

The actual data is not exported
- internal admins
- tokens
- token owners/assignments
- subscriptions
- audit
- stats

**Comment by cornelinux** [comment]
There are several requirements, we need to take into account:

## Req1: Easy integration into pi-manage

The logic within pi-manage should be minimal and the integration should be easy. 
We need to avoid the necessity to change pi-manage with new configs popping up.

pi-manage should not need to know, which config sections exist. Rather it should be able to query the server for the available config sections like

~~~~python
# fetch a list of config sections or classes
available_config_sections = get_config_sections()
~~~~

This should be used in e.g. displaying help information. The admin could then choose which sections (or all) he wants to import or export.

~~~~python
for config_section in selected_config_sections:
    # process each section that was selected by admin
    _foo = config_section.export_config()
~~~~

This way we would achieve that
* we have no internal logic in pi-manage
* we do not need to change pi-manage if the server gets new config sections.

## Req2: Introduce minimal code changes

We need to introduce minimal code changes to assure stability within version 3 for our users.

Config data is stored in many different ways within privacyIDEA for several reasons. Since some configuration contain logic and used everywhere (like policies) and other config data does not contain any logic and is only used on demand (like privacyIDEA servers or PROXY config, if we would implement this)

We should add **class methods** for exporting and importing the data. Since each class knows how data is handled within. If it is stored in one object (policies) or in several objects (smtpserver).
Also: Not all objects exist at runtime! I.e. data needs to be read from the database anyways.

Using two class methods also allow to not only export and import config data, but any data.

We could also use class methods on internal admins, token classes, audit data... everything.

## Req3: Easy adding of new configs

Adding config sections should be done where it logic happens, i.e. within the (config) classes.
This way we avoid to edit several locations whenever we add a new exportable data.
(Maybe we should talk about exportable data anyways instead of export/import config)

This could be done by registering classes or exportes, similar to the way we do this with registering events:
https://github.com/privacyidea/privacyidea/blob/master/privacyidea/api/validate.py#L157

We should decorate the class (or the method), that is capabable of exporting/importing data. 
This way, when adding a new config/data section we simply need to add the decorated export/import methods in only one place and do nothing else.

The additional upside is, that as soon as we have introduced this framework, config/data sections can be added one by one.
So the whole task could be easily split in setting up the framework and then addind configs one by one.

## Side notes

* This way we could even define an exporter/importer for any tokenclass, if we wanted to.
* The export /import methods should take care of additional aspects like: when importing, should old data be orverwritten (deleted) or should the imported data be "added" to the existing config.

**Comment by plettich** [comment]
I have started a concept page here: https://github.com/privacyidea/privacyidea/wiki/concept:-configuration-import-export

**Comment by cornelinux** [comment]
I am moving this to 3.7. We implemented the basic functionality for 3.6 and we can add further sections in upcoming releases.

**Comment by laclaro** [comment]
There are basically two ways how we list configuration: 
- as a dictionary with identifiers as keys and the configurations data as values like e.g. for sms gateways.
- as e.g. for sms providers or as a list of dictionaries, where the identifier is within the dictionary as for policies

One could unify these to have less individual code. Also for some components we deliver more or different information on listing as we need for saving.
